### PR TITLE
Streamline foreign stored context restore and allow to perserve response headers

### PR DESCRIPTION
**Pull request by s1monw**
Today we do not preserve response headers if they are present on a transport protocol
response. While preserving these headers is not always desired, in the most cases we
should pass on these headers to have consistent results for depreciation headers etc.
yet, this hasn't been much of a problem since most of the deprecations are detected early
ie. on the coordinating node such that this bug wasn't uncovered until #22647

This commit allow to optionally preserve headers when a context is restored and also streamlines
the context restore since it leaked frequently into the callers thread context when the callers
context wasn't restored again.

### CODE REVIEW

**Review by jaymode — APPROVED** [review]
LGTM. this really streamlines and cleans things up nicely

**Comment by s1monw** [comment]
@elasticmachine test this please

File: core/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 1]
```diff
@@ -377,12 +377,15 @@ void retry(Exception failure) {
                 onFailure(failure);
                 return;
             }
-            final ThreadContext.StoredContext context = threadPool.getThreadContext().newStoredContext();
+            final ThreadContext threadContext = threadPool.getThreadContext();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: nit: threadContext is not used

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 2]
```diff
@@ -514,24 +514,28 @@ public void onFailure(Exception e) {
                             request),
                     e);
                 request.onRetry();
-                final ThreadContext.StoredContext context = threadPool.getThreadContext().newStoredContext();
+                final Supplier<ThreadContext.StoredContext> context = threadPool.getThreadContext().newRestorableContext(false);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: Is this needed? I think the observer takes care of this? If not, I can work on folding this into the observer as a follow up.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 3]
```diff
@@ -809,12 +813,13 @@ void retry(Exception failure) {
             }
             setPhase(task, "waiting_for_retry");
             request.onRetry();
-            final ThreadContext.StoredContext context = threadPool.getThreadContext().newStoredContext();
+            final Supplier<ThreadContext.StoredContext> context = threadPool.getThreadContext().newRestorableContext(false);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: same comment about the observer.

File: core/src/test/java/org/elasticsearch/common/util/concurrent/ThreadContextTests.java [hunk 4]
```diff
@@ -109,6 +110,57 @@ public void testStoreContext() {
         assertNull(threadContext.getHeader("foo.bar"));
     }
 
+    public void testRestorableContext() {
+        Settings build = Settings.builder().put("request.headers.default", "1").build();
+        ThreadContext threadContext = new ThreadContext(build);
+        threadContext.putHeader("foo", "bar");
+        threadContext.putTransient("ctx.foo", 1);
+        Supplier<ThreadContext.StoredContext> contextSupplier = threadContext.newRestorableContext(true);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/common/util/concurrent/ThreadContextTests.java]: shall we add a duplicate response header too? I.e., one that is both in the original context and the one being replcaed?

**Review by bleskes — CHANGES_REQUESTED** [review]
Beautiful - left a single question about the observer usage

**Comment by s1monw** [comment]
@bleskes I pushed changes

**Review by bleskes — APPROVED** [review]
Thx @s1monw

**Pull request merged**